src/main.ts

```diff
@@ -129,7 +136,15 @@ async function run(): Promise<void> {
     const cargo = await rustCore.Cargo.get();
 
     await installCargoSemverChecks(cargo);
+
+    const cache = new RustdocCache(cargo);
+    const cacheFound = await cache.restore();
+
     await runCargoSemverChecks(cargo);
+
+    if (!cacheFound) {
+        await cache.save();
+    }
```

Comment:
Well, maybe I didn't understand caching correctly. Is it possible that new crates will be added during running `cargo-semver-checks`? Since we hash all `Cargo.lock` files and use the hash as a part of the cache key, the cache won't be restored if any of them changes.

Reply:
Hmm, are we guaranteed that newly added packages always cause a change in `Cargo.lock` even if they only use the same dependencies as other crates in the workspace? Will that continue to be the case forever, even if the `Cargo.lock` format changes? It seems to be on version 3 right now, so it could have a version 4 in the future.

I'd recommend looking into what `cache.save()` does if the cached data hasn't changed, and how long that takes. It's possible that the savings of not calling the function are minimal, and would be immediately wiped out (and then some) if we ever have to debug a problem related to this.

I trust you to look into the data and then use your best judgment on what to do here. You're the expert on this, having spent much more time looking into this than I have, so I trust you to make a good decision 👍

Reply:
`saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them. It's a very quick process, and even for large cache it would be a little fraction of the whole running time, so removing this `if` is not a problem. 

However, I still can't understand why we would need to change it, but this is probably my misunderstanding of how caching works in `cargo-semver-checks`. Could you explain it a little more? What do you mean by newly added packages? And can you give some concrete example of a situation when this approach might fail?

Reply:
No problem, here's one scenario.

Say there's a workspace that contains multiple crates, and uses this action. Then, a new PR adds a new crate. This crate has no _new_ dependencies that weren't previously already used in the workspace — perhaps it has no dependencies _whatsoever_.

Does this modify the lockfile? In today's v3 lockfile format, yes: an entry for the new crate will be added to the lockfile.

_Must it_ modify the lockfile? No — no new information needed to be added, since this case could have been an "implicit default" case in the lockfile format: "if a crate doesn't have an entry in the lockfile, all its dependencies listed in `Cargo.toml` take the versions already specified in the lockfile." A hypothetical future lockfile format could add this behavior, since AFAIK the lockfile format doesn't ever promise "new packages added to the workspace will cause an edit to the lockfile." (There's also [Hyrum's Law](https://www.hyrumslaw.com/) to consider, and I'm sure the Rust team is aware of that law so it probably means this change won't actually be made...)

In that case, the hash and cache key would still match and will correctly _partially_ populate the cache — but the action won't upload the cached baseline for the _newly added crate in the PR_. Only when a new cargo-semver-checks version or rustc version is released will this situation get resolved, since that will reset the cache key.

This scenario isn't particularly likely (mostly because of the Rust team's competence and Hyrum's Law) so maybe we don't have to worry about it. But it's important to realize it's a possibility, seriously consider, and then make an explicit decision on ("worry about it / don't worry about it") rather than never thinking about it and making an unintentional random choice.

Reply:
To be clear: it's _your call_ whether we leave the `if` or take it out. I was trying to help flesh out the pros and cons so you can have accurate input data to use when making the decision.

Reply:
Thank you for such a detailed response! I was pretty sure that if a workspace contains multiple crates, a `Cargo.lock` is created separately for each of them, which is obviously wrong! Therefore I couldn't understand how the hash might not change if one new file is being hashed... Now I see that the difference is much more subtle, so it's better to save the cache again in any case.

However, there's one thing that is still bothering me. We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?

Reply:
> `saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them.

I'm not sure if this is true: based on the [code here](https://github.com/actions/cache/blob/main/src/saveImpl.ts#L40-L49) it seems like the action will recompute the cache key and will not upload if the key didn't change.

That means we have to include the cached rustdoc JSON files in the cache key.

> We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?

Good point. Two things we should strongly consider in this case:
- Add a `prefix-key` argument like in [the Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) to allow users to manually flush the cache if needed.
- Ensure that if the `Cargo.lock` file didn't exist _before_ the action, then it isn't taken into account when computing the cache key _after_ the action. This is because `Cargo.lock` will be created as part of most `cargo` operations in a workspace that doesn't contain it. We don't want those newly-generated `Cargo.lock` files going into the cache-saving cache key calculation because then we're guaranteeing that we never get any cache hits.

Reply:
> > `saveCache` provided by `@actions/cache` doesn't check whether the cached files changed, just uploads them.
> 
> I'm not sure if this is true: based on the [code here](https://github.com/actions/cache/blob/main/src/saveImpl.ts#L40-L49) it seems like the action will recompute the cache key and will not upload if the key didn't change.
> That means we have to include the cached rustdoc JSON files in the cache key.

This is the code of the GitHub `cache` action, which might be used in workflows, but not from the TypeScript code. Please notice that it calls `saveCache` function from the `@actions/cache` module (https://github.com/actions/cache/blob/main/src/saveImpl.ts#L59-L64) and this is the same function I'm using in `RustdocCache`. The code of `saveCache` can be found here https://github.com/actions/toolkit/blob/main/packages/cache/src/cache.ts#L165, and I'm pretty convinced it doesn't compare the key with the previous one, just uploads the files. Moreover, we cannot include the hash of rustdoc files in the key, because then we won't know the key at the beginning of the action and won't be able to download the cache... But we can just remove the `if` leaving bare `await cache.save()` and know the cache will change when it should. 

However I think we can do a bit better then that: download the cache, hash the cache directory, then run `cargo-semver-checks` and finally hash the cache directory again. Then we can do `cache.save()` only if these two hashes differ. I think it won't make much difference for small crates, but for these with really big rustdoc hashing the cache twice might be several seconds faster than uploading it unnecessarily. What do you think about it?

> > We're checking semver for library crates, isn't it very rare for them to contain `Cargo.lock` file?
> 
> Good point. Two things we should strongly consider in this case:
> 
>     * Add a `prefix-key` argument like in [the Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) to allow users to manually flush the cache if needed.

Such an input already exists! I've called it `cache-key` and it defaults to `${{ github.job }}`. Do you think it should be described (or named) differently?

>     * Ensure that if the `Cargo.lock` file didn't exist _before_ the action, then it isn't taken into account when computing the cache key _after_ the action. This is because `Cargo.lock` will be created as part of most `cargo` operations in a workspace that doesn't contain it. We don't want those newly-generated `Cargo.lock` files going into the cache-saving cache key calculation because then we're guaranteeing that we never get any cache hits.

This isn't a problem, because the key is computed only once, before running `cargo-semver-checks`.

Reply:
Good catch, I missed the distinction between the actions implementation and the function called here.

> However I think we can do a bit better then that: download the cache, hash the cache directory, then run cargo-semver-checks and finally hash the cache directory again. Then we can do cache.save() only if these two hashes differ.

I think this is a great idea. Rustdoc JSON files can be hundreds of MB in size, and if we can avoid doing expensive compression and network operations with a quick hash check, we should.

Is it easy enough to implement now, or should we open an issue and tackle it in a follow-up PR?

> Such an input already exists! I've called it `cache-key` and it defaults to `${{ github.job }}`.

While I agree that `cache-key` is _capable_ of accomplishing the same thing, I don't think it's equivalent. Its stated intended role is to control which jobs share a cache, not do cache-busting. When used for cache-busting, let's keep in mind that most users won't be nearly as familiar with GitHub Actions as you and I have become, and are likely to make a mistake. For example, instead of changing `cache-key` from `${{ github.job }}` to `${{ github.job }}-1`, they might just change it to `1` and then be frustrated at the fact the cache is now shared across jobs.

In other words, we are not ensuring our users [fall into the pit of success](https://blog.codinghorror.com/falling-into-the-pit-of-success/) (highly recommended short read!)

This is why the [Rust caching action](https://github.com/Swatinem/rust-cache#example-usage) has so many different cache key settings: instead of saying "here's one cache key setting that technically can do everything," it says "here are many settings, each of which does exactly one thing and you can't accidentally misuse it to do anything else.

We should consider adopting a similar design and perhaps even the same naming terminology, with one key for cache-busting, another key for sharing caches across jobs, etc.